models/meme.go

```diff
@@ -2,20 +2,67 @@ package models
 
 import (
 	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
 
 	"github.com/boltdb/bolt"
 )
 
 var bucketName = []byte("memes")
+var binBucketName = []byte("memes_bindata")
 
 // Meme is a meme
 type Meme struct {
-	URL string `json:"url"`
-	ID  uint64 `json:"id"`
+	URL         string `json:"url"`
+	ID          uint64 `json:"id"`
+	OriginalUrl string `json:"original-url"`
+	ContentType string `json:"content-type"`
+	bindata     []byte
+}
+
+func MemeFromBolt(k []byte, v []byte) (*Meme, error) {
+	meme := new(Meme)
+	if err := json.Unmarshal(v, meme); err != nil {
+		return meme, err
+	}
+	meme.ID = binary.BigEndian.Uint64(k)
+	return meme, nil
+}
+
+// Fetch all data from memes bucket and assign to ptr
+func MemeFetch(ptr *Meme, id uint64) func(tx *bolt.Tx) error {
+	return func(tx *bolt.Tx) error {
+		meme, err := MemeFromBolt(itob(id), tx.Bucket(bucketName).Get(itob(id)))
+		if err != nil {
+			log.Fatalln(err)
+			return err
+		}
+		*ptr = *meme
+		return nil
+	}
+}
+
+// Retrieve binary image for a meme (from memes_bindata bucket) and assign
+// to m.bindata
+func (m *Meme) GetImage(tx *bolt.Tx) error {
+	m.bindata = tx.Bucket(binBucketName).Get(itob(m.ID))
+	return nil
+}
+
+// Writa binary image to a given io.Writer
+func (m *Meme) WriteImage(w io.Writer) error {
+	if _, err := w.Write(m.bindata); err != nil {
+		return err
+	}
+	return nil
 }
 
 // Save saves meme in bolt storage
 func (m *Meme) Save(tx *bolt.Tx) error {
+	// TODO: use a proper transaction commit and rollback
```

Comment:
It already is a transaction as `h.db.Update(meme.Save)` wraps the callback in a transaction :)